### ISSUE TITLE
swagger-metadata's processSwaggerDocuments is Sync but contains ASync function

### ISSUE THREAD

**Issue by alasdairhurst** [apigee-127/swagger-tools#491]
When processSwaggerDocuments is called, it tries to resolve the swagger spec using an async function, however, it treats it as a synchronous action: https://github.com/apigee-127/swagger-tools/blob/master/middleware/swagger-metadata.js#L337
There is a comment about this being expected, however the first time that this function is called, there is no cache. This means that further down the line, the swagger will be validated asynchronously, since no cache exists.
https://github.com/apigee-127/swagger-tools/blob/master/lib/specs.js#L1453
This means that when processSwaggerDocuments calls createCacheEntry for each entry, it will be *after* the validation has occurred. But by this time, processSwaggerDocuments has already returned the (empty) cache before it had time to set. Since the cache is an empty object when processSwaggerDocuments returns, the swagger documents will fail to process. 
https://github.com/apigee-127/swagger-tools/blob/master/middleware/swagger-metadata.js#L384
This results in issues when calling the swagger-metadata middleware the first time, since it won't match any routes in the cache

I'm guessing this is something to do with not calling swaggerTools.initializeMiddleware first, but it doesn't seem like any of your tests are actually testing the middleware directly so I can see why you may not have come across it. I can't use initializeMiddleware either since my use case involves loading and managing multiple swagger documents in one service, where an invalid swagger document is logged, bu the server loads the rest. However, any validation failures or errors in initializeMiddleware cause the process to exit. (dependencies should NOT do this!)  It would also mean I need to initializeMiddleware for every single swagger document, creating a new middleware function for each, when calling a single one should work fine, other than this issue with the asynchronous function call inside a synchronous function just because it expects a cache to exist from an unrelated action.

**Comment by whitlockjc** [comment]
The thing with `swagger-tools` is it's a bit opinionated.  There is zero, and I seriously mean **ZERO** benefit to having `swagger-tools` startup if the Swagger document is invalid.  Since validating a Swagger document requires full resolution, which is an asynchronous process, this is why `initializeMiddleware` is **required** _(which is heavily documented mind you)_.

So while I do agree that most middleware should not use `process.exit`, my worry was that people would ignore any errors in the validation and just spin up a server blindly.  Then they'd file bugs and complain that `swagger-tools` isn't working, yada yada, and I'd end up chasing my tail only to find out they decided to start `swagger-tools` with busted Swagger documents.

There's been discussion about removing the hard exist on failure to validate the Swagger document but at least now you know why it was done, and it wasn't simply _not knowing the rules_.

As for not being able to use `initializeMiddleware` for multiple Swagger documents, that's not true.  In fact, we've seen people doing this already.  Chaining async functions is not really hard.

I'm closing this issue because it's not reporting anything new and the chance that the flow will change would be handled in #327.

**Comment by alasdairhurst** [comment]
@whitlockjc  I didn't mean that I couldn't use ```initializeMiddleware``` for multiple swagger docs, but for multiple swagger docs where one or more could be invalid. In this case the application handles any errors with validation and either shuts down cleanly or continues with clearly marked errors in the UI. (It's software for API Creation and Management where the swagger docs are application artifacts rather than just defining the application itself).

I actually looked into this issue and code further, and the issue was partially on my end. The application was adding runtime extensions to the swagger doc after the swagger-tools validation. (based on extensions which swagger tools does not know about) This meant that the cache with a key based on the md5 hash of the doc wasn't getting used when the middleware was called.

At runtime, this wasn't much of a problem. When the server span up, the validation was done, and middleware was generated (causing the cache to get set), however the unit tests which were testing the middleware function were creating the middleware, then calling it straight away. This meant that the middleware creation had returned a function, but it was still calling the async function that was setting the (new) cache when the middleware was called, causing it to not perform as expected (req.swagger not set).
 This is a very specific use case, which was fixed by setting a timeout after creating the middleware, but it's not the best solution. If cache is going to be set at all when the swagger-metadata middleware is requested, there's no point creating the cache after the function returns right?
